vsphere/internal/vmworkflow/virtual_machine_customize_subresource.go

```diff
@@ -0,0 +1,490 @@
+	v4mask := d.Get(netifKey("ipv4_netmask", n)).(int)
+	v4gw, v4gwOk := d.Get(cKeyPrefix + "." + "ipv4_gateway").(string)
+	var obj types.CustomizationIPSettings
+	switch {
+	case v4addrOk:
+		obj.Ip = &types.CustomizationFixedIp{
+			IpAddress: v4addr.(string),
+		}
+		obj.SubnetMask = v4CIDRMaskToDotted(v4mask)
+		// Check for the gateway
+		if v4gwAdd && v4gwOk && matchGateway(v4addr.(string), v4mask, v4gw) {
+			obj.Gateway = []string{v4gw}
+			v4gwFound = true
+		}
+	default:
+		obj.Ip = &types.CustomizationDhcpIpGenerator{}
+	}
+	obj.DnsServerList = structure.SliceInterfacesToStrings(d.Get(netifKey("dns_server_list", n)).([]interface{}))
+	obj.DnsDomain = d.Get(netifKey("dns_domain", n)).(string)
+	obj.IpV6Spec, v6gwFound = expandCustomizationIPSettingsIPV6AddressSpec(d, n, v6gwAdd)
+	return obj, v4gwFound, v6gwFound
+}
+
+// expandSliceOfCustomizationAdapterMapping reads certain ResourceData keys and
+// returns a CustomizationAdapterMapping slice.
+func expandSliceOfCustomizationAdapterMapping(d *schema.ResourceData) []types.CustomizationAdapterMapping {
+	s := d.Get(cKeyPrefix + "." + "network_interface").([]interface{})
+	if len(s) < 1 {
+		return nil
+	}
+	result := make([]types.CustomizationAdapterMapping, len(s))
+	var v4gwFound, v6gwFound bool
+	for i := range s {
+		var adapter types.CustomizationIPSettings
+		adapter, v4gwFound, v6gwFound = expandCustomizationIPSettings(d, i, !v4gwFound, !v6gwFound)
+		obj := types.CustomizationAdapterMapping{
+			Adapter: adapter,
+		}
+		result[i] = obj
+	}
+	return result
+}
+
+// ExpandCustomizationSpec reads certain ResourceData keys and
+// returns a CustomizationSpec.
+func ExpandCustomizationSpec(d *schema.ResourceData, family string) types.CustomizationSpec {
+	obj := types.CustomizationSpec{
+		Identity:         expandBaseCustomizationIdentitySettings(d, family),
+		GlobalIPSettings: expandCustomizationGlobalIPSettings(d),
+		NicSettingMap:    expandSliceOfCustomizationAdapterMapping(d),
+	}
+	return obj
+}
+
+// ValidateCustomizationSpec checks the validity of the supplied customization
+// spec. It should be called during diff customization to veto invalid configs.
+func ValidateCustomizationSpec(d *schema.ResourceDiff, family string) error {
+	// Validate that the proper section exists for OS family suboptions.
+	lExists := len(d.Get(cKeyPrefix+"."+"linux_options").([]interface{})) > 0
+	wExists := len(d.Get(cKeyPrefix+"."+"windows_options").([]interface{})) > 0
+	wrExists := d.Get(cKeyPrefix+"."+"windows_sysprep_text").(string) != ""
+	switch {
+	case family == string(types.VirtualMachineGuestOsFamilyLinuxGuest) && !lExists:
+		return errors.New("linux_options must exist in VM customization options for Linux operating systems")
+	case family == string(types.VirtualMachineGuestOsFamilyWindowsGuest) && !wExists && !wrExists:
+		return errors.New("one of windows_options or windows_sysprep_text must exist in VM customization options for Windows operating systems")
```

Comment:
Your error says that `wExists` or `wrExists` needs to exist but you are checking that both need to exist. `!wExists && !wrExists`. Do you need to specify both or just one?

Reply:
This is a bit confusing at face value, but `&&` is correct here - the error condition should only fire if _both_ are missing. If one of them is there, then that's passable.

Reply:
Ah, yep! I see it now